File: exercises/csharp/Workshop/Exercise3/OrderProcessor.cs
using System;

namespace Workshop.Exercise3;

public class OrderResult
{
    public bool IsSuccess { get; set; }
    public decimal FinalAmount { get; set; }
    public string Message { get; set; } = string.Empty;
}

public class OrderProcessor
{
    public OrderResult ProcessOrder(decimal amount, string state, bool isMember, int loyaltyYears)
    {
        var res = new OrderResult();

        if (amount > 0)
        {
            if (state == "OH" || state == "AZ" || state == "TX")
            {
                decimal tax = 0;
                if (state == "OH")
                {
                    tax = amount * 0.0575m;
                }
                else if (state == "AZ")
                {
                    tax = amount * 0.056m;
                }
                else
                {
                    tax = amount * 0.0625m;
                }

                decimal disc = 0;
                if (isMember)
                {
                    if (loyaltyYears >= 5)
                    {
                        disc = amount * 0.15m;
                    }
                    else if (loyaltyYears >= 2)
                    {
                        disc = amount * 0.05m;
                    }
                    else
                    {
                        disc = amount * 0.02m;
                    }
                }

                decimal calc = amount + tax - disc;
                if (calc < 0)
                {
                    calc = 0;
                }

                res.IsSuccess = true;
                res.FinalAmount = Math.Round(calc, 2);
                res.Message = "Processed";
            }
            else
            {
                res.IsSuccess = false;
                res.FinalAmount = 0;
                res.Message = "Unsupported State";
            }
        }
        else
        {
            res.IsSuccess = false;
            res.FinalAmount = 0;
            res.Message = "Invalid Amount";
        }

        return res;
    }
}

File: exercises/java/src/main/java/workshop/exercise3/OrderProcessor.java
package workshop.exercise3;

import java.math.BigDecimal;
import java.math.RoundingMode;

public class OrderProcessor {
    public OrderResult processOrder(BigDecimal amount, String state, boolean isMember, int loyaltyYears) {
        if (amount != null && amount.compareTo(BigDecimal.ZERO) > 0) {
            if ("OH".equals(state) || "AZ".equals(state) || "TX".equals(state)) {
                BigDecimal tax;
                if ("OH".equals(state)) {
                    tax = amount.multiply(new BigDecimal("0.0575"));
                } else if ("AZ".equals(state)) {
                    tax = amount.multiply(new BigDecimal("0.056"));
                } else {
                    tax = amount.multiply(new BigDecimal("0.0625"));
                }

                BigDecimal disc = BigDecimal.ZERO;
                if (isMember) {
                    if (loyaltyYears >= 5) {
                        disc = amount.multiply(new BigDecimal("0.15"));
                    } else if (loyaltyYears >= 2) {
                        disc = amount.multiply(new BigDecimal("0.05"));
                    } else {
                        disc = amount.multiply(new BigDecimal("0.02"));
                    }
                }

                BigDecimal calc = amount.add(tax).subtract(disc);
                if (calc.compareTo(BigDecimal.ZERO) < 0) {
                    calc = BigDecimal.ZERO;
                }

                BigDecimal finalAmount = calc.setScale(2, RoundingMode.HALF_UP);
                return new OrderResult(true, finalAmount, "Processed");
            } else {
                return new OrderResult(false, BigDecimal.ZERO, "Unsupported State");
            }
        } else {
            return new OrderResult(false, BigDecimal.ZERO, "Invalid Amount");
        }
    }
}
